### DISCUSSION TITLE
Batch resizing while keeping the original

### DISCUSSION THREAD

**Discussion by ANewGithubNoob** [ImageMagick/ImageMagick#5267] [Help] [answered]
Hey guys. I really don't get any further. I want to resize multiple images like flower.jpg, house.jpg into smaller versions with size attached. Like flower_600x480.jpg, house_600x435.jpg. The size conversion and naming works as planned, but the original file ends up the smallest size and I have the feeling I always resize the image before, which results in quality loss over time compared to decreasing quality and resizing from the original.

Here is my script so far:
```
magick mogrify 
-set filename:name %t -write '%[filename:name]_original.jpg' *.jpg 
-resize 800 -quality 50% -set filename:name %t -set filename:size %wx%h -write '%[filename:name]_%[filename:size].jpg' *.jpg 
-resize 600 -quality 50% -set filename:name %t -set filename:size %wx%h -write '%[filename:name]_%[filename:size].jpg' *.jpg 
-resize 400 -quality 50% -set filename:name %t -set filename:size %wx%h -write '%[filename:name]_%[filename:size].jpg' *.jpg 
-resize 225 -quality 50% -set filename:name %t -set filename:size %wx%h -write '%[filename:name]_%[filename:size].jpg' *.jpg
```
I want every line to operate with the original source (if the original stays original, I no longer need the _original backup).

PS: parameters like resize and quality may change, so I'd rather keep it that long and easy changeable.

Could you please help me?

**Comment by fmw42** [comment]
I do not think you can use -set filename:name and then -write to output specific filenames in magick mogrify.  I believe you will need to write a script loop using magick (not magick mogrify) to process your images. See https://imagemagick.org/Usage/files/#write

**Comment by snibgo** [comment]
As fmw42 says.

I wouldn't attempt this with "magick mogrify". Instead, I would use a simple "magick" command, on a single file. When that works as I want, I would put that in a shell loop.

**Comment by GeeMack** [comment, marked as answer]
As others have mentioned, "mogrify" probably isn't the best ImageMagick tool for this task. This would usually be done by processing multiple files, one at a time, in a shell loop with a "magick" command.

A single IMv7 command to create multiple outputs of various dimensions from a single input file, and with formatted file names, might look like this.

     set IMG=Sun Halo 001.jpg
     
     magick "%IMG%" -set filename:n "%t_original" ^
        ( -clone 0 -resize 800 -set filename:n "%t_%wx%h" ) ^
        ( -clone 0 -resize 600 -set filename:n "%t_%wx%h" ) ^
        ( -clone 0 -resize 400 -set filename:n "%t_%wx%h" ) ^
        ( -clone 0 -resize 225 -set filename:n "%t_%wx%h" ) ^
        -quality 50% "%[filename:n].jpg"

This command is in Windows CMD syntax, and works on Windows 10 with IMv7. The method you use to put it in a loop depends on your OS, command line or script, and it may require escaping parentheses or changing quote marks, continued-line character, etc.

Inside each set of parentheses the original input is cloned and resized, and given the appropriate output file name. An interesting feature of setting the IM "filename:n" variable for each resize, after the resize operations the command writes all five output JPG images, each with it's own appropriate file name.

**Reply by ANewGithubNoob** [reply]
1. I don't exactly understand the very first line / the "%IMG%" after magick. Could I just use something like *.jpg instead of listing every single image?
I just want all pictures in the folder to be looped.

2. Is it easy to convert this to powershell? I can easily call powershell in every folder I'm in, so that would be a bit easier. But no big problem, if not.

3. I guess I can set quality within each paranthese if I want to reduce quality differently in different sizes?

Thank you

**Reply by GeeMack** [reply]
The %IMG% is the variable that holds the name of the input image in Windows cmd or BAT script.  That would be set to each filename at the top of a loop, or in my example the first line sets the input image to `Sun Halo 001.jpg`. 

I don't know Powershell, but the example command should work in a Windows BAT script by just doubling the percent signs "%%" everywhere except Windows variables like "%IMG%".

I don't think you can set the output quality separately for each individual image the way you can with the filenames. You could clone the original, resize it, set the filename, set the quality, and write the output all inside each set of parentheses.

**Reply by snibgo** [reply]
As GeeMack says.

> Could I just use something like *.jpg instead of listing every single image?

No. If you have more than one input file, "-clone 0" won't do what you need.
